src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationAnalysisValueProvider.cs

```diff
@@ -0,0 +1,50 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+
+namespace Microsoft.CodeAnalysis.Diagnostics
+{
+    /// <summary>
+    /// Wrapper over the core <see cref="AnalysisValueProvider{TKey, TValue}"/> which holds a strong reference to key-value pairs for the lifetime of a compilation that this provider is associated with.
+    /// This ensures that values are never re-computed for equivalent keys while analyzing each compilation, improving overall analyzer performance.
+    /// </summary>
+    internal sealed class CompilationAnalysisValueProvider<TKey, TValue>
+        where TKey : class
+    {
+        private readonly AnalysisValueProvider<TKey, TValue> _analysisValueProvider;
+        private readonly Dictionary<TKey, TValue> _valueMap;
+
+        public CompilationAnalysisValueProvider(AnalysisValueProvider<TKey, TValue> analysisValueProvider)
+        {
+            _analysisValueProvider = analysisValueProvider;
+            _valueMap = new Dictionary<TKey, TValue>(analysisValueProvider.KeyComparer);
+        }
+
+        internal bool TryGetValue(TKey key, out TValue value)
+        {
+            // First try to get the cached value for this compilation.
+            lock (_valueMap)
```

Comment:
I'm having trouble figuring out what the two disconnected lock statements guarantee. They don't prevent multiple threads from performing the same computation. If that does occur, they don't guarantee that any particular computed value is what gets stored in the map.

They do prevent simultaneously querying and updating the map. Is that the sole intent?

A comment explaining the purpose of the locking would help.